internal/rbd/rbd_attach.go

```diff
@@ -193,35 +204,46 @@ func waitForPath(ctx context.Context, pool, namespace, image string, maxRetries
 	return "", false
 }
 
-// Check if rbd-nbd tools are installed.
-func checkRbdNbdTools() bool {
+// getRbdNbdToolsVersion returns rbd-nbd tools version.
+func getRbdNbdToolsVersion() string {
 	// check if the module is loaded or compiled in
 	_, err := os.Stat(fmt.Sprintf("/sys/module/%s", moduleNbd))
 	if os.IsNotExist(err) {
 		// try to load the module
 		_, _, err = util.ExecCommand(context.TODO(), "modprobe", moduleNbd)
 		if err != nil {
-			log.ExtendedLogMsg("rbd-nbd: nbd modprobe failed with error %v", err)
+			log.WarningLogMsg("rbd-nbd: nbd modprobe failed with error %v", err)
 
-			return false
+			return ""
 		}
 	}
-	if _, _, err := util.ExecCommand(context.TODO(), rbdTonbd, "--version"); err != nil {
-		log.ExtendedLogMsg("rbd-nbd: running rbd-nbd --version failed with error %v", err)
+	stdout, stderr, err := util.ExecCommand(context.TODO(), rbdTonbd, "--version")
+	if err != nil || stderr != "" {
+		log.WarningLogMsg("rbd-nbd: running rbd-nbd --version failed with error:%v, stderr:%s", err, stderr)
+
+		return ""
+	}
+	ver := strings.Split(stdout, " ")
+	// ex: ceph version 15.2.14 cd3bb7e87a2f62c1b862ff3fd8b1eec13391a5be) octopus (stable)
+	rbdNbdVer := ver[2]
+
+	if rbdNbdVer == "" {
+		log.WarningLogMsg("rbd-nbd: --version command didn't return version, stdout:%s", stdout)
 
-		return false
+		return ""
 	}
-	log.ExtendedLogMsg("rbd-nbd tools were found.")
 
-	return true
+	log.DefaultLog("rbd-nbd tools were found with version: %s", rbdNbdVer)
+
+	return rbdNbdVer
 }
 
 func attachRBDImage(ctx context.Context, volOptions *rbdVolume, device string, cr *util.Credentials) (string, error) {
 	var err error
 
 	image := volOptions.RbdImageName
 	useNBD := false
-	if volOptions.Mounter == rbdTonbd && hasNBD {
+	if volOptions.Mounter == rbdTonbd {
```

Comment:
we no longer default to krbd if nbd is not present?

Reply:
Not sure on this part, but how does it matter for us here (just curious)?

Reply:
if the nbd module was not available on the node we use krbd to map the image by default with this change we don't default to anything we always try to use nbd if it's specified in storageclass?

Reply:
> this change we don't default to anything we always try to use nbd if it's specified in storageclass?
I'm sorry, I might be missing something, which part of the code changes from this PR is leading to the behaviour.

Reply:
`if volOptions.Mounter == rbdTonbd && hasNBD {` to `if volOptions.Mounter == rbdTonbd`

Reply:
this is only to avoid checking if rbd-nbd is installed or not (which is redundant) rest remain as it is IMHO.

Reply:
Previously there NBD was only used when `hasNBD` was set. So, in the case where `checkRbdNbdTools()` returned false, a StorageClass configured to use NBD would fallback to krbd instead. This is a change in behaviour, and we do not know if there are any users that rely on it.

Reply:
right!
so the question is
1. should we have `nbdLoaded` variable that checks if the module is loaded, if not fall back to `krbd` ?
2. Or just fail if the storage class asks for `rbd-nbd` and `lsmod` doesn't list it?

We can still avoid hasNBD checking for rbd-nbd binary availability on the container which is redundant.

Reply:
@nixpanic @Madhu-1 check the latest updates, I have left hasNBD for module loaded case, and added `hasNBDCookieSupport` separately. Hope that will keep the old style of fallback intact.

Thanks!